File: CleanCodeJava/app/src/main/java/io/chuumong/cleancodejava/data/api/util/ThreadCallAdapter.java
package io.chuumong.cleancodejava.data.api.util;

import java.lang.reflect.Type;

import retrofit2.Call;
import retrofit2.CallAdapter;
import rx.Observable;

/**
 * Created by JongHunLee on 2017-02-22.
 */

class ThreadCallAdapter implements CallAdapter<Observable<?>> {

    private final CallAdapter<?> adapter;
    private final Scheduler scheduler;

    ThreadCallAdapter(CallAdapter<?> adapter, Scheduler scheduler) {
        this.adapter = adapter;
        this.scheduler = scheduler;
    }

    @Override
    public Type responseType() {
        return adapter.responseType();
    }

    @Override
    public <R> Observable<?> adapt(Call<R> call) {
        if (adapter == null) {
            return null;
        }

        Object adapt = adapter.adapt(call);

        // subscribeOn 및 observeOn 할당
        if (adapt instanceof Observable<?>) {
            return ((Observable) adapt).subscribeOn(scheduler.background()).observeOn(scheduler.main());
        }

        return null;
    }
}

File: CleanCodeKotlin/app/src/main/java/io/chuumong/cleancodekotlin/data/api/util/ThreadCallAdapter.kt
package io.chuumong.cleancodekotlin.data.api.util

import retrofit2.Call
import retrofit2.CallAdapter
import rx.Observable
import java.lang.reflect.Type

/**
 * Created by JongHunLee on 2017-02-23.
 */
class ThreadCallAdapter(val adapter: CallAdapter<*>?, val scheduler: Scheduler) : CallAdapter<Observable<*>> {
    override fun <R : Any?> adapt(call: Call<R>?): Observable<*>? {
        if (adapter == null) {
            return null
        }

        val adapt = adapter.adapt(call)
        if (adapt is Observable<*>) {
            return adapt.subscribeOn(scheduler.background()).observeOn(scheduler.main())
        }

        return null
    }

    override fun responseType() = adapter?.responseType()
}